beacon-chain/sync/data_column_sidecars.go

```diff
@@ -0,0 +1,854 @@
+// for a given root only if its head state is higher than the block slot.
+func computeIndicesByRootByPeer(
+	p2p prysmP2P.P2P,
+	slotByBlockRoot map[[fieldparams.RootLength]byte]primitives.Slot,
+	indicesByBlockRoot map[[fieldparams.RootLength]byte]map[uint64]bool,
+	peers map[goPeer.ID]bool,
+) (map[goPeer.ID]map[[fieldparams.RootLength]byte]map[uint64]bool, error) {
+	// First, compute custody columns for all peers
+	peersByIndex := make(map[uint64]map[goPeer.ID]bool)
+	headSlotByPeer := make(map[goPeer.ID]primitives.Slot)
+	for peer := range peers {
+		// Computes the custody columns for each peer
+		nodeID, err := prysmP2P.ConvertPeerIDToNodeID(peer)
+		if err != nil {
+			return nil, errors.Wrapf(err, "convert peer ID to node ID for peer %s", peer)
+		}
+
+		custodyGroupCount := p2p.CustodyGroupCountFromPeer(peer)
+
+		dasInfo, _, err := peerdas.Info(nodeID, custodyGroupCount)
+		if err != nil {
+			return nil, errors.Wrapf(err, "peerdas info for peer %s", peer)
+		}
+
+		for column := range dasInfo.CustodyColumns {
+			if _, exists := peersByIndex[column]; !exists {
+				peersByIndex[column] = make(map[goPeer.ID]bool)
+			}
+			peersByIndex[column][peer] = true
+		}
+
+		// Compute the head slot for each peer
+		peerChainState, err := p2p.Peers().ChainState(peer)
+		if err != nil {
+			return nil, errors.Wrapf(err, "get chain state for peer %s", peer)
+		}
+
+		if peerChainState == nil {
+			return nil, errors.Errorf("chain state is nil for peer %s", peer)
+		}
+
+		headSlotByPeer[peer] = peerChainState.HeadSlot
+	}
+
+	// For each block root and its indices, find suitable peers
+	indicesByRootByPeer := make(map[goPeer.ID]map[[fieldparams.RootLength]byte]map[uint64]bool)
+	for blockRoot, indices := range indicesByBlockRoot {
+		blockSlot, ok := slotByBlockRoot[blockRoot]
+		if !ok {
+			return nil, errors.Errorf("slot not found for block root %#x", blockRoot)
+		}
+
+		for index := range indices {
+			peers := peersByIndex[index]
+			for peer := range peers {
+				peerHeadSlot, ok := headSlotByPeer[peer]
+				if !ok {
+					return nil, errors.Errorf("head slot not found for peer %s", peer)
+				}
+
+				if peerHeadSlot < blockSlot {
+					continue
+				}
+
+				// Build peers->root->indices map
+				if _, exists := indicesByRootByPeer[peer]; !exists {
+					indicesByRootByPeer[peer] = make(map[[fieldparams.RootLength]byte]map[uint64]bool)
+				}
+				if _, exists := indicesByRootByPeer[peer][blockRoot]; !exists {
+					indicesByRootByPeer[peer][blockRoot] = make(map[uint64]bool)
+				}
+				indicesByRootByPeer[peer][blockRoot][index] = true
+			}
+		}
+	}
+
+	return indicesByRootByPeer, nil
+}
+
+// randomPeer selects a random peer. If no peers has enough bandwidth, it will wait and retry.
+// Returns the selected peer ID and any error.
+func randomPeer(
+	ctx context.Context,
+	randomSource *rand.Rand,
+	rateLimiter *leakybucket.Collector,
+	count int,
+	indicesByRootByPeer map[goPeer.ID]map[[32]byte]map[uint64]bool,
```

Comment:
small nit I think i've seen you use fieldparams.RootLength in most places we should just replace 32 with fieldparams.RootLength

Reply:
Thanks!

Fixed in 09b7fa6bc9a0da746069b96e0a40fa8d3cfd0a33.